### PR TITLE
Some CI setup simplification and debugging

### PR DESCRIPTION
**Pull request by ararslan**
Changes:
- Use version numbers for actions instead of the commit SHA for the version's git tag. The SHA makes it impossible to tell whether an action is outdated, and might interfere with how dependabot works but that's just a guess.
- Update versions of Julia-related actions
- Switch from actions/cache to julia-actions/cache
- I forget whether I made any other changes

The S3 upload step failed for the most recent manually-triggered run because the download-artifact action couldn't find what the upload-artifact action had (successfully!) uploaded. I have no idea why that'd happen but the changes here are in an attempt to debug.

### CODE REVIEW

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -35,27 +37,13 @@ jobs:
         os:
           - ubuntu-latest
     steps:
-      - uses: actions/checkout@5a4ac9002d0be2fb38bd78e4b4dbde5606d7042f
-
-      - uses: julia-actions/setup-julia@0b9b1d2cd24245f151902702d8e73b3f6b910014
+      - uses: actions/checkout@v4
```
Comment (DilumAluthge) [inline .github/workflows/CI.yml:40]: I'd recommend doing this:
```suggestion
      - uses: actions/checkout@d632683dd7b4114ad314bca15554477dd762a938 # v4.2.0
```

This way, you get the reproducibility of the commit hash, but you still have the human-readable version in the comment.

Dependabot will update both the commit hash and also the comment with the version number.

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -68,46 +56,28 @@ jobs:
...
         run: npx -p ajv-cli@3.3.0 ajv -s schema.json -d versions.json
-
       - name: Upload versions.json as workflow artifact
-        uses: actions/upload-artifact@c24449f33cd45d4826c6702db7e49f7cdb9b551d
+        uses: actions/upload-artifact@v4
         with:
           name: versions
           path: versions.json
           if-no-files-found: error
+          overwrite: true
```
Comment (DilumAluthge) [inline .github/workflows/CI.yml:80]: Why is this change necessary?

Comment (ararslan) [reply .github/workflows/CI.yml:80]: Probably isn't, was for debugging

File: .github/workflows/CI.yml (hunk 1 above)
Comment (ararslan) [reply .github/workflows/CI.yml:40]: We can keep commit hashes if you want but since we're only using some very well-established actions I'm not worried about malicious actors or semver violations, which IIUC are really the only reasons why one would want to use commit hashes. IMO the tradeoff for brevity/readability here seems worthwhile.

Comment (DilumAluthge) [reply .github/workflows/CI.yml:40]: I don't feel particular strongly either way. We could do `action@v1.2.3`. That gives us the ability to easily roll back if something breaks, and is less verbose than `action@commithash # v1.2.3`. Dependabot works with any of these formats.

**Comment by ararslan** [comment]
@DilumAluthge, can we try this out? I had forgotten about this PR and ran the workflow to update for 1.11.0 then it failed on S3 upload. 🫠 I can work around it again but it'd be nice if it actually worked lol

**Comment by DilumAluthge** [comment]
Yeah let's revive this.

For the actions versions, for now, let's go with `action@v1.2.3`? We can always switch to `actions@commithash # v1.2.3` later if we want. I'm not a huge fan of `action@v1`, because it's not as reproducible.

And also revert that above change that we might not need?

File: .github/workflows/CI.yml (hunk 2 above)
Comment (DilumAluthge) [reply .github/workflows/CI.yml:80]: Let's try reverting this change for now, and see if everything still works?

**Comment by ararslan** [comment]
Actually, now that dependabot is updating workflow versions with the existing commit hashes, we might as well just merge its open PRs, since they cover most of what's in here. The only functional-ish change this makes otherwise is swapping actions/cache for julia-actions/cache.

**Comment by DilumAluthge** [comment]
Cool. Let me first merge the GitHub Merge Queue PR. Then we can merge all the Dependabot PRs (merge queue will automatically group them together). Then we can rebase this PR, where the only remaining change will be the `{actions--->julia-actions}/cache` change, as you mentioned.

**Comment by DilumAluthge** [comment]
Closing in favor of #23
